lib/Onyx.js

```diff
@@ -867,9 +857,25 @@ function set(key, value) {
     cache.set(key, value);
     notifySubscribersOnNextTick(key, value);
 
-    // Write the thing to persistent storage, which will trigger a storage event for any other tabs open on this domain
-    return Storage.setItem(key, value)
-        .catch(error => evictStorageAndRetry(error, set, key, value));
+    return Promise.resolve();
+}
+
+/**
+ * Write a value to our store with the given key
+ *
+ * @param {String} key ONYXKEY to set
+ * @param {*} value value to store
+ *
+ * @returns {Promise}
+ */
+function set(key, value) {
+    // Logging properties only since values could be sensitive things we don't want to log
+    Logger.logInfo(`set() called for key: ${key}${_.isObject(value) ? ` properties: ${_.keys(value).join(',')}` : ''}`);
+
+    return new Promise(resolve => Storage.setItem(key, value)
+        .then(updatedValue => resolve(
+            notifyAfterWrite(key, updatedValue, Storage.mergeItem),
+        )).catch(error => evictStorageAndRetry(error, set, key, value)));
```

Comment:
Not sure if I am reading this correctly - but we are using the returned promise value of `Storage.setItem()` and resolving this outer promise with the result of calling `notifyAfterWrite()`. What is the overall thought process behind this change?

Reply:
Doesn't this mean that subscribers will be notified only after we save to storage? My assumption just based on looking at this is that it will be slower to notify subscribers and maybe break some other expectations about the order of operations. I also generally feel like the order of operations here should by "cache", "notify" then "store". That ordering just _feels_ correct to me. Doesn't matter when the writes happen and we are optimizing updating the cache and subscribers.

Reply:
From what you're correct @marcaaron, this would only notify subscribers after writes complete
This is something that we don't want, not just because of slow performance, but also because the write might not complete at all 

We shouldn't wait to notify, but preserve the existing functionality and notify then write

_One of the big performance hits on Android was waiting for the write to complete and then notifying connections_

Reply:
> If i understand the current approach correctly this would essentially mean that we effectively can't use `JSON_PATCH()` on native without manually merging large JS objects too. (because we need to notify before write)

Do we have a proven problem with merging large objects in JS 
Letting native do the merge, would mean we still get one hell of a chunk back - the whole merged object communicated between native and JS 
I think ideally we'd just pass the delta to native, while JS should be performant enough to merge the result to cache

Another way to look at this might be - If large JS objects are a problem - don't have large JS objects - structure App in smaller chunks (paging for report actions)

Reply:
I'm not sure if I follow the discussion completely, but my thinking is:

1. if we are calling `Onyx.set()` there is no advantage to JSON_PATCH()
2. if we are calling `Onyx.merge()` then we should still "merge the complete object" in the JS cache. But when we do the hand off to storage just pass the partial object to SQLite JSON_PATCH() if we can (native) otherwise (web) set the complete object.

Reply:
>if we explicitly want sqlite to handle merging directly on native, instead of merging big objects in onyx before passing the new data to the storage provide

said another way, I don't see a clear way for this to work atm without major changes so I would suggest not trying to avoid it and just focus on implementing JSON_PATCH() for partial objects. The "cache merge" and the "storage merge" behavior should match so it seems fine to pass partial objects to SQLite.